### PR TITLE
Neural net estimator

### PR DESCRIPTION
**Pull request by abolshov**
start code for Mass Estimator based on neural net

### CODE REVIEW

File: Studies/NeuralNetEstimator/Common/__init__.py
Comment (kandrosov) [inline Studies/NeuralNetEstimator/Common/__init__.py:1]: add `__init__.py` to `.gitignore`

File: Studies/NeuralNetEstimator/Reco/create_config.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import io 
+import yaml
+
+cfg = {
```
Comment (kandrosov) [inline Studies/NeuralNetEstimator/Reco/create_config.py:4]: what is the purpose of this script? Why not create a `yaml` config by hand and commit it instead?

File: Studies/NeuralNetEstimator/Reco/test_JetNet.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+from Common.JetNet import JetNet
+from Common.JetNetData import JetNetData
+from Common.JetNet_utils import PlotPrediction
+import yaml 
+
+
+def main():
+    jet_obs = ['px', 'py', 'pz', 'E', 'btagPNetB', 'PNetRegPtRawCorr', 'PNetRegPtRawCorrNeutrino', 'PNetRegPtRawRes']
+    features = [f'centralJet{i}_{var}' for i in range(10) for var in jet_obs]
+    labels = ['H_VV_px', 'H_VV_py', 'H_VV_pz', 'H_VV_E', 'X_mass']
```
Comment (kandrosov) [inline Studies/NeuralNetEstimator/Reco/test_JetNet.py:10]: take `jet_obs`, `labels` and `nJets` from config

File: Studies/NeuralNetEstimator/Reco/test_JetNet.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+    input_files = ["test_GluGlutotoRadiontoHHto2B2Vto2B2JLNu_450.root"]
```
Comment (kandrosov) [inline Studies/NeuralNetEstimator/Reco/test_JetNet.py:12]: provide list of input files as command line arguments

File: Studies/NeuralNetEstimator/Reco/test_JetNet.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+        data = JetNetData(features, labels)
+        data.ReadFiles(input_files)
+
+        net = JetNet(features, labels, cfg)
+        # net.LoadModel("./models/JetNet_v2.keras")
+        net.LoadModel("./models/nn_estimator_v1.keras")
+
+        pred = net.Predict(data.data[features])
+        mass = PlotPrediction(pred, data.data)
```
Comment (kandrosov) [inline Studies/NeuralNetEstimator/Reco/test_JetNet.py:25]: pass model file name and where to store outputs in the command line

File: Studies/Purity/ReportUtils.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import csv
```
Comment (kandrosov) [inline Studies/Purity/ReportUtils.py:1]: all purity related files shouldn't be part of this PR

File: Studies/NeuralNetEstimator/Reco/train_JetNet.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        data = JetNetData(features, labels)
+        data.ReadFiles(input_files)
+        
+        data.Shuffle()
+        data.TrainTestSplit()        
+
+        net = JetNet(features, labels, cfg)
+        net.ConfigureModel(data.train_features.shape)
+        history = net.Fit(data.train_features, data.train_labels)
+        net.SaveModel("./models/")
```
Comment (kandrosov) [inline Studies/NeuralNetEstimator/Reco/train_JetNet.py:26]: same as for test: provide parameters in config, inputs & outputs in the command line

File: Studies/NeuralNetEstimator/Common/JetNet.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        self.verbosity = cfg['verbosity']
+        self.valid_split = cfg['valid_split']
+        self.name = cfg['name']
+        self.topology = cfg['topology']
+
+        self.model = None
+
+
+    def ConfigureModel(self, dataset_shape):
+        self.model = tf.keras.Sequential([tf.keras.layers.Dense(layer_size, activation='relu') for layer_size in self.topology])
```
Comment (kandrosov) [inline Studies/NeuralNetEstimator/Common/JetNet.py:27]: Define the model into a separate file and load it dynamically (based on what is specified in the config)

File: Studies/NeuralNetEstimator/Common/JetNet.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
+import tensorflow as tf
+import pandas as pd
+import numpy as np
+import uproot
+
+from Common.JetNet_utils import MXLossFunc, H_WW_MXLossFunc
+
+
+class JetNet():
```
Comment (kandrosov) [inline Studies/NeuralNetEstimator/Common/JetNet.py:9]:
```suggestion
class JetNetFitter:
```

File: Studies/NeuralNetEstimator/Common/JetNetData.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    def ReadFiles(self, input_files):
+        for file_name in input_files:
+            self.ReadFile(file_name)
+
+
+    def Shuffle(self):
+        self.data = shuffle(self.data)
+
+
+    def TrainTestSplit(self):
```
Comment (kandrosov) [inline Studies/NeuralNetEstimator/Common/JetNetData.py:48]: implement `SelectEvents(self, values, mod_param, is_train)` and apply it before shuffle. Use `event % mod_param == value` to select events to make the selection reproducible.

File: Studies/NeuralNetEstimator/Common/JetNetData.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        # pd.DataFrame containing full dataset
+        self.data = pd.DataFrame(columns=[*features, *labels])
+
+        self.train_frac = train_test_fraction
+
+        self.train_features = pd.DataFrame()
+        self.test_features = pd.DataFrame()
+
+        self.train_labels = pd.DataFrame()
+        self.test_labels = pd.DataFrame()
```
Comment (kandrosov) [inline Studies/NeuralNetEstimator/Common/JetNetData.py:24]:
```suggestion
```

File: Studies/NeuralNetEstimator/Common/JetNet_utils.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
+import tensorflow as tf
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+@tf.function
+def MXLossFunc(target, output):
+    H_WW_px = target[:, 0]
+    H_WW_py = target[:, 1]
+    H_WW_pz = target[:, 2]
```
Comment (kandrosov) [inline Studies/NeuralNetEstimator/Common/JetNet_utils.py:10]:
```suggestion
    H_WW_p3 = target[:, 0:3]
```

File: Studies/NeuralNetEstimator/Common/JetNet_utils.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    H_WW_E = target[:, 3]    
+        
+    X_mass = target[:, 4]
+    
+    H_bb_px = output[:, 0]
+    H_bb_py = output[:, 1]
+    H_bb_pz = output[:, 2]
```
Comment (kandrosov) [inline Studies/NeuralNetEstimator/Common/JetNet_utils.py:17]:
```suggestion
    H_bb_p3 = output[:, 0:3]
```

File: Studies/NeuralNetEstimator/Common/JetNet_utils.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    H_mass = 125.0
+    H_bb_E_sqr = H_mass*H_mass + H_bb_px*H_bb_px + H_bb_py*H_bb_py + H_bb_pz*H_bb_pz
```
Comment (kandrosov) [inline Studies/NeuralNetEstimator/Common/JetNet_utils.py:19]:
```suggestion
    H_bb_E_sqr = H_mass*H_mass + tf.sum(H_bb_p3 ** 2, axis=1)
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Studies/.gitignore [hunk 14]
```diff
@@ -0,0 +1 @@
+Studies/NeuralNetEstimator/Common/__init__.py
```
Comment (kandrosov) [inline Studies/.gitignore:1]:
```suggestion
__init__.py
```

**Review by kandrosov — CHANGES_REQUESTED** [review]
